### ISSUE TITLE
Crashes without error when executing user32.SetWindowLongPtrW() with a WindowProc Callback

### ISSUE THREAD

**Issue by Spookyfan** [Koromix/koffi#128]
Trying to migrate an application from ffi-napi and everything works fine except for `user32.SetWindowLongPtrW()`
When it executes it crashes without any error message.
I try to set an Electron created window as a taskbar for Windows with that

Here are some code snippets

```typescript
// type defs
const HANDLE = koffi.pointer(koffi.opaque('HANDLE'))
const HWND = koffi.alias('HWND', HANDLE)
const UINT = koffi.alias('UINT', 'uint')
const WPARAM = koffi.alias('WPARAM', 'uintptr')
const LPARAM = koffi.alias('LPARAM', 'longlong')
const WNDPROC = koffi.proto('WindowProc', LRESULT, [HWND, UINT, WPARAM, LPARAM])

const SetWindowLongPtrW = user32.func('SetWindowLongPtrW', koffi.pointer(WNDPROC), [HWND, INT, koffi.pointer(WNDPROC)])
```

```typescript
const registerWinPosChangeHook = (windowHandle: bigint, posChangeCallback: (browserWindowHandle: bigint) => void) => {
  let origWindowProc: unknown

  // DModel types are leftover from old lib, but they all map to number or bigint anyway
  const hookCallback = (hWnd: DModel.HWND, msg: DModel.UINT, wParam: DModel.WPARAM, lParam: DModel.LPARAM) => {
    switch (msg) {
      case WM_MOVE:
      case WM_WINDOWPOSCHANGING:
        posChangeCallback(windowHandle)
        return 1
      default:
        // call default WindowProc in case we don't have any custom logic for it.
        const windowProc = user32.CallWindowProcW(origWindowProc, hWnd, msg, wParam, lParam)
        return windowProc
    }
  }

  const callback = koffi.register(hookCallback, koffi.pointer(WNDPROC))

  origWindowProc = user32.SetWindowLongPtrW(windowHandle, GWLP_WNDPROC, callback)
}
```
`posChangeCallback`  executes `setTaskbarPos()` 
```typescript
// hWnd belongs to a newly created window that we created in a previous function with SHAppBarMessage(ABM_QUERYPOS, appBarData)
const setTaskbarPos = (hWnd: bigint) => {
  const appBarData = {
    cbSize: 0,
    hWnd: hWnd,
    uEdge: 0x3,
    rc: {
      left: 0,
      top: height - actualTaskbarHeight,
      right: width,
      bottom: height,
    },
  }

  shell32.SHAppBarMessage(ABM_QUERYPOS, appBarData)
  shell32.SHAppBarMessage(ABM_SETPOS, appBarData)
  user32.SetWindowPos(appBarData.hWnd, HWND_TOP, 0, taskbarBottom, taskbarWidth, actualTaskbarHeight, SWP_NOACTIVATE | SWP_FRAMECHANGED | SWP_NOSENDCHANGING)
}
```

If I try to get an error with `kernel32.getLastError()` it gives me a 203 ERROR_ENVVAR_NOT_FOUND after I execute `user32.SetWindowPos()`

`setTaskbarPos()` on its own works fine without any crashes or errors and also does what it's supposed to do, but inside the WindowProc Callback it crashes, can't figure out what's wrong, so maybe I missed something or it's a Koffi issue or the typing for `user32.SetWindowLongPtrW` and/or `user32.CallWindowProcW` are wrong?

**Comment by Koromix** [comment]
Win32 callbacks typically use the non-default __stdcall convention on x86. If you get it wrong, you'll get a crash in Node.js 32-bit, maybe that's it?

```js
const WNDPROC = koffi.proto('__stdcall', 'WindowProc', LRESULT, [HWND, UINT, WPARAM, LPARAM])
```

Also, this callback will be called after the initial SetWindowLongPtrW() call has returned so you need to register it, as documented here:

https://koffi.dev/callbacks#registered-callbacks

**Comment by Spookyfan** [comment]
I already tried to delcare the functions with __stdcall, but that also didn't work out, besides that I'm using a 64bit system and also compile it to just 64bit

I also registered the callback like that, but it still doesn't work
```typescript
const callback = koffi.register(hookCallback, koffi.pointer(WNDPROC))
```

**Comment by estoykov** [comment]
I have the same issue, the error is:

`[10808:0202/174347.734:ERROR:node_bindings.cc(158)] Fatal error in V8: v8::HandleScope::CreateHandle() Cannot create a handle without a HandleScope
`
Electron: 28.2.1
Koffi: 2.7.3

I was able to successfully use EnumWindows callback though.
And this also work (change and immediately return old wndProc):
```
  baseWndProc = SetWindowProc(handle, GWL_WNDPROC, wndProcCallback);
  SetWindowProc(handle, GWL_WNDPROC, baseWndProc);
```
i.e. I suppose that the method is invoked and the wndProc is correctly set.

I've attached the sample repo - basically the vanilla Electron Filddle repo with subclass related stuff.
[koffi-subclass.zip](https://github.com/Koromix/koffi/files/14142399/koffi-subclass.zip)

**Comment by Koromix** [comment]
I've made some changes that should fix this, at least for @estoykov, but if it's the same problem that'll fix both.

I've published Koffi 2.7.4-beta.1, install with `npm install koffi@beta` and test. Please report to me if it works and I'll make a proper release.

**Comment by Koromix** [comment]
(sorry, closed by mistake)

**Comment by estoykov** [comment]
It's working now ;)
`    "koffi": "^2.7.4-beta.1"`

And btw, sorry, but I forgot to include page loading in the repo:

```
  mainWindow.loadFile('index.html')

  const handle = mainWindow.getNativeWindowHandle().readInt32LE();

```

Anyway, thank you for the quick response!
I'm working with Windows API for almost 30 years, but never imagined that I will have an ability to subclass a window procedure at so "high level" :)

Cheers and keep the good work!

**Comment by Koromix** [comment]
Thank you for the report!

I've made a sample based on the snippets of @Spookyfan, and it seems to work too.

```js
const koffi = require('koffi');
const { app, BrowserWindow } = require('electron');

const HANDLE = koffi.pointer('HANDLE', koffi.opaque());
const HWND = koffi.alias('HWND', HANDLE);
const WNDPROC = koffi.proto('__stdcall', 'WindowProc', 'intptr_t', [HWND, 'uint', 'uintptr_t', 'intptr_t']);

const GWLP_WNDPROC = -4;
const WM_MOVE = 0x0003;
const WM_WINDOWPOSCHANGING = 0x0046;

const user32 = koffi.load('user32.dll');
const SetWindowLongPtrW = user32.func('SetWindowLongPtrW', koffi.pointer(WNDPROC), [HWND, 'int', koffi.pointer(WNDPROC)]);
const CallWindowProcW = user32.func('CallWindowProcW', 'intptr_t', [koffi.pointer(WNDPROC), HWND, 'uint', 'uintptr_t', 'intptr_t']);

const registerWinPosChangeHook = (hWnd, posChangeCallback) => {
    let origWindowProc;

    const hookCallback = (hWnd, msg, wParam, lParam) => {
        switch (msg) {
            case WM_MOVE: {
                let x = lParam & 0xFFFF;
                let y = (lParam >> 16) & 0xFFFF;

                posChangeCallback(hWnd, x, y);
            } break;
        }

        const windowProc = CallWindowProcW(origWindowProc, hWnd, msg, wParam, lParam);
        return windowProc;
    };

    const callback = koffi.register(hookCallback, koffi.pointer(WNDPROC));
    origWindowProc = SetWindowLongPtrW(hWnd, GWLP_WNDPROC, callback);
}

function createWindow () {
    const mainWindow = new BrowserWindow({
        width: 800,
        height: 600
    });

    const hWnd = mainWindow.getNativeWindowHandle().readInt32LE();

    registerWinPosChangeHook(hWnd, (hWnd, x, y) => {
        let msg = `Moving: ${x}x${y}`;
        console.log(msg);
    });
}

app.whenReady().then(() => {
    createWindow();
});

app.on('window-all-closed', function () {
    if (process.platform !== 'darwin')
        app.quit();
});
```

So I've published Koffi 2.7.4 with this fix and a couple other minor things.
